lib/Reader/readerir.cpp

```diff
@@ -3783,6 +3802,248 @@ IRNode *GenIR::genArrayElemAddress(IRNode *Array, IRNode *Index,
+  } else {
+    assert(!ElementTy->isStructTy());
+    makeStoreNonNull(Value, ElementAddress, IsVolatile);
+  }
+
+  return true;
+};
+
+bool GenIR::arrayAddress(CORINFO_SIG_INFO *Sig, IRNode **RetVal) {
+  uint32_t Rank = 0;
+  CorInfoType ElemCorType = CORINFO_TYPE_UNDEF;
+  Type *ElementTy = nullptr;
+  const bool IsStore = false;
+  const bool IsLoadAddr = true;
+
+  if (!canExpandMDArrayRef(Sig, IsStore, IsLoadAddr, &Rank, &ElemCorType,
+                           &ElementTy)) {
+    return false;
+  }
+
+  *RetVal = mdArrayRefAddr(Rank, ElementTy);
+  return true;
+};
+
+bool GenIR::canExpandMDArrayRef(CORINFO_SIG_INFO *Sig, bool IsStore,
+                                bool IsLoadAddr, uint32_t *Rank,
+                                CorInfoType *ElemCorType,
+                                llvm::Type **ElemType) {
+  // TODO: legacy jit limits the number of array intrinsics expanded to 50 to
+  // avoid excessive code expansion. Evaluate whether we need such a limit.
+
+  // If it's a store there is an argument for the value
+  // (doesn't count toward the rank).
+  *Rank = IsStore ? Sig->numArgs - 1 : Sig->numArgs;
+
+  // Early out if we don't want to handle it.
+  // NOTE: rank 1 requires special handling (not done here).
+  if (*Rank > ArrayIntrinMaxRank || *Rank <= 1) {
+    return false;
+  }
+
+  // Figure out the element type
+  CorInfoType CorType = CORINFO_TYPE_UNDEF;
+  CORINFO_CLASS_HANDLE Class = nullptr;
+  uint32_t ElemSize = 0;
+  if (IsStore) {
+    CORINFO_ARG_LIST_HANDLE ArgLst = Sig->args;
+    // Skip arguments for each dimension.
+    for (uint32_t R = 0; R < *Rank; R++) {
+      ArgLst = getArgNext(ArgLst);
+    }
+    CORINFO_CLASS_HANDLE ArgType;
+    CorType = strip(getArgType(Sig, ArgLst, &ArgType));
+
+    ASSERTNR(CorType != CORINFO_TYPE_VAR); // common generics trouble
+
+    if (CorType == CORINFO_TYPE_CLASS || CorType == CORINFO_TYPE_VALUECLASS) {
+      Class = getArgClass(Sig, ArgLst);
+    } else if (CorType == CORINFO_TYPE_REFANY) {
+      Class = getBuiltinClass(CLASSID_TYPED_BYREF);
+    } else {
+      Class = nullptr;
+    }
+  } else {
+    CorType = Sig->retType;
+    Class = Sig->retTypeClass;
+  }
+
+  // Bail if the element type is a value class.
+  if (CorType == CORINFO_TYPE_VALUECLASS) {
+    return false;
+  }
+
+  if (IsStore || IsLoadAddr) {
+    CORINFO_CLASS_HANDLE GCClass = Class;
+    CorInfoType GCCorType = CorType;
+
+    // If it is BYREF then we need to find the child type
+    // for the next security check.
+    if (CorType == CORINFO_TYPE_BYREF) {
+      CORINFO_CLASS_HANDLE ClassTmp = nullptr;
+      GCCorType = getChildType(Class, &ClassTmp);
+      GCClass = ClassTmp;
+    }
+
+    // See if the element is an object.
+    if (GCCorType == CORINFO_TYPE_STRING || GCCorType == CORINFO_TYPE_CLASS) {
+      // If it's not final, we can't expand due to security reasons.
+      uint32_t ClassAttribs = getClassAttribs(GCClass);
```

Comment:
Yes. ArrayTypeMismatchException should be thrown in a case like this:
class B {}
class C : B {}
class D : B {}
C[,] arrayC = new C[2, 2];
B[,] arrayB = arrayC;
arrayB[0, 0] = new D();

If B is final (i.e., sealed) this is not possible.